### PR TITLE
Implement ETH Shanghai upgrade 

### PR DESCRIPTION
**Pull request by libotony**
# Description

This PR adds a fork temporally named `ETH_SH` including the following VIPs:

+ [VIP-242: ETH Shanghai Upgrade Specification](https://github.com/vechain/VIPs/blob/master/vips/VIP-242.md)
 
## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] New and existing E2E tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules
- [x] I have not added any vulnerable dependencies to my code

### CODE REVIEW

**Review by darrenvechain — APPROVED** [review]

**Review by darrenvechain — DISMISSED** [review]

File: vm/opcodes.go [hunk 1]
```diff
@@ -25,11 +25,7 @@ type OpCode byte
 
 // IsPush specifies if an opcode is a PUSH opcode.
 func (op OpCode) IsPush() bool {
-	switch op {
-	case PUSH1, PUSH2, PUSH3, PUSH4, PUSH5, PUSH6, PUSH7, PUSH8, PUSH9, PUSH10, PUSH11, PUSH12, PUSH13, PUSH14, PUSH15, PUSH16, PUSH17, PUSH18, PUSH19, PUSH20, PUSH21, PUSH22, PUSH23, PUSH24, PUSH25, PUSH26, PUSH27, PUSH28, PUSH29, PUSH30, PUSH31, PUSH32:
```
Comment (otherview) [inline vm/opcodes.go:29]: Curious way of checking..

**Review by otherview — APPROVED** [review]

File: vm/jump_table_test.go [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+package vm
+
+import (
+	"testing"
+
+	"github.com/holiman/uint256"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestPush0OpCode(t *testing.T) {
```
Comment (leszek-vechain) [inline vm/jump_table_test.go:15]: can we add similar test for BASEFEE ?

File: thor/fork_config.go [hunk 3]
```diff
@@ -19,6 +19,7 @@ type ForkConfig struct {
 	ETH_IST   uint32
 	VIP214    uint32
 	FINALITY  uint32
+	ETH_SH    uint32
```
[the remark this answers is not in the document]
Comment (libotony) [reply thor/fork_config.go:22]: It was temporally named ETH_SH, we can rename it when the name is finalized.

File: genesis/customnet_test.go [hunk 4]
```diff
@@ -136,7 +136,7 @@ func TestNewCustomGenesisMarshalUnmarshal(t *testing.T) {
...
...00000000000000000000000000000000000000000000000000006d6173746572"},{"masterAddress":"0x361277d1b27504f36a3b33d3a52d1f8270331b8c","endorsorAddress":"0x361277d1b27504f36a3b33d3a52d1f8270331b8c","identity":"0x00000000000000000000000000000000000000000000000000006d6173746572"},{"masterAddress":"0xd7f75a0a1287ab2916848909c8531a0ea9412800","endorsorAddress":"0xd7f75a0a1287ab2916848909c8531a0ea9412800","identity":"0x00000000000000000000000000000000000000000000000000006d6173746572"},{"masterAddress":"0xabef6032b9176c186f6bf984f548bda53349f70a","endorsorAddress":"0xabef6032b9176c186f6bf984f548bda53349f70a","identity":"0x00000000000000000000000000000000000000000000000000006d6173746572"},{"masterAddress":"0x865306084235bf804c8bba8a8d56890940ca8f0b","endorsorAddress":"0x865306084235bf804c8bba8a8d56890940ca8f0b","identity":"0x00000000000000000000000000000000000000000000000000006d6173746572"}],"params":{"rewardRatio":"-0x64","baseGasPrice":"0x0","proposerEndorsement":"0x0","executorAddress":null,"maxBlockProposers":10000},"executor":{"approvers":null},"forkConfig":{"VIP191":4294967295,"ETH_CONST":4294967295,"BLOCKLIST":4294967295,"ETH_IST":4294967295,"VIP214":4294967295,"FINALITY":0,"ETH_SH":0}}`
```
[the remark this answers is not in the document]
Comment (libotony) [reply genesis/customnet_test.go:139]: For testing perspective, would be better to have both activated and default value. I think it has nothing to do with `acting ETH_SH in customnet`, wdyt?

File: vm/contracts.go [hunk 5]
```diff
@@ -70,34 +70,62 @@ var PrecompiledContractsIstanbul = map[common.Address]PrecompiledContract{
...
+	for k := range PrecompiledContractsShanghai {
+		PrecompiledAddressesShanghai = append(PrecompiledAddressesShanghai, k)
+	}
 }
 
 // ActivePrecompiles returns the precompiles enabled with the current configuration.
 func ActivePrecompiles(rules Rules) []common.Address {
 	switch {
+	case rules.IsShanghai:
+		return PrecompiledAddressesShanghai
```
[the remark this answers is not in the document]
Comment (libotony) [reply vm/contracts.go:122]: You're right, we started from byzantium, no need to involve homestead,I might just copied the function from upsteam.

File: vm/jump_table_test.go (hunk 2 above)
Comment (libotony) [reply vm/jump_table_test.go:15]: BASEFEE is tested in runtime package, this file was aligned with upstream.

File: genesis/customnet_test.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (otherview) [reply genesis/customnet_test.go:139]: Yeah that's fair. And what do you think about adding the default `ETH_SH: 0` to CustomNetWithParams ?

**Review by otherview — APPROVED** [review]

File: genesis/customnet_test.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (libotony) [reply genesis/customnet_test.go:139]: I see your point, will update accordingly

**Review by darrenvechain — APPROVED** [review]

**Comment by libotony** [comment]
Closing since the feature has been re-scheduled.

**Pull request closed without merging**